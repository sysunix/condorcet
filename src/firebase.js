import firebase from "@firebase/app";
import "@firebase/firestore";
import "@firebase/auth";
import "@firebase/functions";

const {
  VUE_APP_API_KEY,
  VUE_APP_AUTH_DOMAIN,
  VUE_APP_DATABASE_URL,
  VUE_APP_PROJECT_ID,
  VUE_APP_STORAGE_BUCKET,
  VUE_APP_MESSAGING_SENDER_ID
} = process.env;

const config = {
  apiKey: VUE_APP_API_KEY,
  authDomain: VUE_APP_AUTH_DOMAIN,
  databaseURL: VUE_APP_DATABASE_URL,
  projectId: VUE_APP_PROJECT_ID,
  storageBucket: VUE_APP_STORAGE_BUCKET,
  messagingSenderId: VUE_APP_MESSAGING_SENDER_ID
};

firebase.initializeApp(config);

export const db = firebase.firestore();
export const auth = firebase.auth();
export const functions = firebase.functions();

export default firebase;
