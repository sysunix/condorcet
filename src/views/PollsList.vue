<template>
  <Loader :processing="processing">
    <Empty
      :condition="userPolls.length === 0"
      :text="$t('empty.userPolls.text')"
      :action-text="$t('empty.userPolls.action')"
      to="/polls/new"
    />

    <div class="flex">
      <div class="flex flex-wrap w-full">
        <div
          v-for="poll in userPolls"
          :key="poll.id"
          class="w-full md:w-1/2 lg:w-1/3 py-2"
        >
          <Card
            class="md:mx-2"
            v-bind="poll"
            :user-id="userId"
            @onShare="copyMagicLink"
            @onToggleStatus="togglePoll"
            @onDelete="deletePoll"
            @onLeave="leavePoll"
          />
        </div>
      </div>
    </div>
  </Loader>
</template>

<script>
import { mapActions, mapGetters } from "vuex";
import { db } from "../firebase";
import Card from "../components/Card";
import Loader from "../components/Loader.vue";
import Empty from "../components/Empty";

export default {
  name: "PollsList",
  components: {
    Card,
    Loader,
    Empty
  },
  computed: {
    ...mapGetters("user", ["userId"]),
    ...mapGetters("poll", ["userPolls"])
  },
  mounted() {
    this.listenPolls(this.userId);
  },
  methods: {
    ...mapActions("app", ["addNotification"]),
    ...mapActions("poll", ["listenPolls", "removePoll"]),
    async togglePoll(id) {
      try {
        const document = db.collection("polls").doc(id);
        const currentPollState = (await document.get()).data();

        const newPollState = {
          ...currentPollState,
          isActive: !currentPollState.isActive
        };

        document.set(newPollState);
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème`,
          status: "error"
        });
      }
    },
    async leavePoll(id) {
      if (confirm("Tu veux quitter ce scrutin ?") === false) {
        return;
      }

      const poll = (await db
        .collection("polls")
        .doc(id)
        .get()).data();

      const users = poll.users.filter(user => user !== this.userId);

      db.collection("polls")
        .doc(id)
        .set({ ...poll, users });

      this.removePoll(id);

      this.addNotification({
        text: `Tu as bien quitté le scrutin ${poll.question}`,
        status: "info"
      });
    },
    deletePoll(id) {
      if (confirm("Confirmer le suppression ?") === false) {
        return;
      }

      try {
        db.collection("polls")
          .doc(id)
          .delete();

        this.removePoll(id);

        this.addNotification({
          text: "Le scrutin  a bien été supprimé",
          status: "info"
        });
      } catch (error) {
        this.addNotification({
          text: `Il y a eu un problème lors de la suppression de ${name}`,
          status: "error"
        });
      }
    },
    async copyMagicLink(id) {
      const poll = this.userPolls.find(poll => poll.id === id);
      const link = `${window.location.origin}/polls/${poll.id}/join?token=${
        poll.token
      }`;

      try {
        await this.$copyText(link);
        this.addNotification({
          text: `Lien copié`,
          status: "info"
        });
      } catch (error) {
        this.addNotification({
          text: `Lien copié`,
          status: "error"
        });
      }
    }
  }
};
</script>

<style></style>
