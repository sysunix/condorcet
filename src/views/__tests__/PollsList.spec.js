import { shallowMount, createLocalVue } from "@vue/test-utils";
import Vuex from "vuex";

import { db } from "../../firebase";
import {
  MOCK_USER_ID,
  MOCK_POLL_ID,
  MOCK_POLL,
  getTranslation
} from "../../utils/test";
import PollsList from "../PollsList.vue";

const localVue = createLocalVue();
localVue.use(Vuex);

jest.mock("../../firebase");

const $copyText = jest.fn();

describe("PollsList.vue", () => {
  let wrapper;
  let appActions;
  let pollActions;
  let pollGetters;

  beforeEach(() => {
    jest.clearAllMocks();

    pollActions = {
      listenPolls: jest.fn(),
      leavePoll: jest.fn(),
      removePoll: jest.fn()
    };

    appActions = {
      addNotification: jest.fn()
    };

    pollGetters = {
      userPolls: jest.fn(() => [{ id: MOCK_POLL_ID, ...MOCK_POLL }])
    };

    const store = new Vuex.Store({
      modules: {
        app: {
          namespaced: true,
          actions: appActions
        },
        user: {
          namespaced: true,
          getters: {
            userId: () => MOCK_USER_ID
          }
        },
        poll: {
          namespaced: true,
          actions: pollActions,
          getters: pollGetters
        }
      }
    });

    wrapper = shallowMount(PollsList, {
      store,
      stubs: ["router-link"],
      localVue,
      mocks: { $copyText, $t: getTranslation }
    });
  });

  it("should render with message saying there is no polls", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render with one poll", () => {
    pollGetters.userPolls;
    expect(wrapper).toMatchSnapshot();
  });

  it("should leave poll", () => {
    window.confirm = jest.fn(() => true);
    wrapper.vm.leavePoll(MOCK_POLL_ID);
    expect(db.collection).toBeCalledWith("polls");
    expect(db.collection("polls").doc).toBeCalledWith(MOCK_POLL_ID);
  });

  it("should not leave poll", () => {
    window.confirm = jest.fn(() => false);
    wrapper.vm.leavePoll(MOCK_POLL_ID);
    expect(db.collection).not.toBeCalled();
    expect(db.collection("polls").doc).not.toBeCalled();
  });

  it("should copy magic link", async () => {
    await wrapper.vm.copyMagicLink(MOCK_POLL_ID);

    expect(wrapper.vm.$copyText).toBeCalled();
  });
});
