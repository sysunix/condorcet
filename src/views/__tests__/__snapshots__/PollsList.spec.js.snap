// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PollsList.vue should render with message saying there is no polls 1`] = `
<loader-stub>
  <empty-stub text="Tu n'a pas encore de scrutin." actiontext="Créer un scrutin" to="/polls/new"></empty-stub>
  <div class="flex">
    <div class="flex flex-wrap w-full">
      <div class="w-full md:w-1/2 lg:w-1/3 py-2">
        <card-stub question="Qui est le meilleur ?" description="Qui repellendus est aliquam vel qui." id="MOCK_POLL_ID" userid="MOCK_USER_ID" owner="MOCK_USER_ID" isactive="true" users="MOCK_USER_ID" answers="[object Object],[object Object],[object Object]" ispublic="true" timestamp="[object Object]" token="MOCK_TOKEN" class="md:mx-2"></card-stub>
      </div>
    </div>
  </div>
</loader-stub>
`;

exports[`PollsList.vue should render with one poll 1`] = `
<loader-stub>
  <empty-stub text="Tu n'a pas encore de scrutin." actiontext="Créer un scrutin" to="/polls/new"></empty-stub>
  <div class="flex">
    <div class="flex flex-wrap w-full">
      <div class="w-full md:w-1/2 lg:w-1/3 py-2">
        <card-stub question="Qui est le meilleur ?" description="Qui repellendus est aliquam vel qui." id="MOCK_POLL_ID" userid="MOCK_USER_ID" owner="MOCK_USER_ID" isactive="true" users="MOCK_USER_ID" answers="[object Object],[object Object],[object Object]" ispublic="true" timestamp="[object Object]" token="MOCK_TOKEN" class="md:mx-2"></card-stub>
      </div>
    </div>
  </div>
</loader-stub>
`;
